ARCs/arc-0200.md

```diff
@@ -0,0 +1,226 @@
+---
+arc: 200
+title: Algorand Smart Contract Token Specification
+description: Base specification for tokens implemented as smart contracts
+author: Nicholas Shellabarger (@temptemp3)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/223
+status: Draft
+type: Standards Track
+category: Interface
+created: 2023-07-03
+requires: 3, 4, 22, 28
+---
+
+# Algorand Smart Contract Token Specification
+
+## Abstract
+
+This specifies an interface for tokens to be implemented on Algorand as smart contracts.
+This interface defines a minimal interface for tokens to be held and transferred, to be augmented by other standard interfaces and custom methods.
+
+
+## Motivation
+
+Currently, most tokens in the Algorand ecosystem are implemented as ASAs.
+However, to provide rich extra functionality, it can be desirable to implement tokens as a smart contract instead.
+To foster an interoperable token ecosystem, it is necessary that the core interfaces for tokens be standardized.
+
+
+## Specification
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+
+### Core Token specification
+
+A smart contract token that is compliant with this standard MUST implement the following interface:
+
+```json
+{
+  "name": "ARC-200",
+  "desc": "Smart Contract Token Base Interface",
+  "methods": [
+    {
+      "name": "name",
+      "desc": "Returns the name of the token",
+      "readonly": true,
+      "args": [],
+      "returns": { "type": "bytes[32]", "desc": "The name of the token" }
+    },
+    {
+      "name": "symbol",
+      "desc": "Returns the symbol of the token",
+      "readonly": true,
+      "args": [],
+      "returns": { "type": "bytes[8]", "desc": "The symbol of the token" }
+    },
+    {
+      "name": "decimals",
```

Comment:
The reason for treating `decimals` as an individual field is similar to how `name` or `symbol` are handled as separate fields. I prefer not to use `ufixed` encoded decimal because it would limit the implementation to a specific encoding method. The ABI specification only requires `decimals` to be returned as a uint8, but it allows the smart contract to store the decimal value in any desired format. The choice of implementation is left to the developer.

One possible implementation approach is to validate that `decimals` falls within an acceptable range. For compatibility with ASAs, the decimals value could be checked to be less than 19. For ERC20 compatibility, an implementation would check that it is less than 256.

Additionally, based on my understanding, Reach does not utilize `ufixed`. If a value is known to be less than 256 at compile-time, the compiler would automatically use uint8.

Considering the potential inconveniences and the availability of alternative methods, I don't believe using `ufixed` is necessary to achieve the desired purpose.

Reply:
Yeah the more I think about it the more it makes sense to keep it as a separate method since clients will need an easy way to discover how many decimals are actually supported regardless of encoding. 

I will say though that I think the only language that should dictate the direction of standards is TEAL. Standards should dictate language features, not the other way around. Obviously it is extra dev effort and will take longer to implement/support across the ecosystem. But long term, having a standard that everyone feels is best suited for its purpose is better than having multiple evolving standards superseding each other.

Reply:
Here since this values may use UInt256, `uint8` may be preferable `decimals`

Reply:
> Yeah the more I think about it the more it makes sense to keep it as a separate method since clients will need an easy way to discover how many decimals are actually supported regardless of encoding.

Yup. It is an individual field. Disregard my tangent regarding decimals and uint8. Similar to using uint256 for totalSupply, balances, and allowance values, the decimals method should return a uint8.

The rest are implementation details which I hope we all agree is out of the scope of the spec.